include/seqan3/core/detail/pack_algorithm.hpp

```diff
@@ -0,0 +1,244 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides algorithms for meta programming, parameter packs and seqan3::type_list.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <seqan3/core/type_list/type_list.hpp>
+#include <seqan3/core/tuple_utility.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/type_traits>
+
+namespace seqan3::detail
+{
+
+//!\cond
+template <typename type_list_t>
+struct type_list_expander;
+//!\endcond
+
+/*!\brief Helper class to invoke a meta algorithm on the types contained in a seqan3::type_list.
+ * \tparam type_list_t The type list given as a template template parameter.
+ * \tparam args_t The template arguments contained in `type_list_t` to apply the target function on.
+ *
+ * \details
+ *
+ * The meta algorithms provide a parameter pack version and a seqan3::type_list version.
+ * In the type list version the algorithm is called on the types contained in the enclosing type list after they have
+ * been wrapped in std::type_identity. Thus, the type list version uses the parameter pack version to call the passed
+ * function on the types.
+ *
+ * Concrete this helper does the following:
+ *  * expand the types within the type list
+ *  * wrap each type in std::identity and instantiate it (i.e. transform a type into a value)
+ *  * call a target function with the instances of std::identity (as a pack).
+ *
+ * This is a technical trick to make a type representable as a value. Instantiating a type might not always work
+ * because not every type provides a default constructor. In addition it is possible to use incomplete types as well.
+ */
+template <template <typename ...> typename type_list_t, typename ...args_t>
```

Comment:
Well, this is an internal data structure that is used by the more or less public interfaces that check the concepts already. So in order to keep compile time to a minimum and since I believe that this structure does not benefit in the regular use case from this concept check, I left it away.